torch_xla/core/xla_builder.py

```diff
@@ -888,91 +947,84 @@ def abstractify(a):  # make a pytree leaf abstract
-    return XlaComputation(name, computation, pick_tensor_args, out_spec)
-
-
-def _jax_to_xla_computation_cache_get_or_insert(jax_func,
-                                                sample_inputs: tuple[Any, ...],
-                                                input_tree_spec,
-                                                get_xla_computation):
-  from jax._src import config
-  global _JAX_TO_XLA_COMPUTATION_CACHE
-  # Use three layers of dictionary lookup.
-  # The first layer uses the `config.trace_context()`, which is strongly referenced.
-  # The second layer uses the `jax_func`, which is weakly referenced.
-  # The third layer uses the sample inputs and the tree spec, which is strongly referenced.
-  config_context_dict = _JAX_TO_XLA_COMPUTATION_CACHE.setdefault(
-      config.trace_context(), WeakKeyDictionary())
-  inner_dict = config_context_dict.setdefault(jax_func, {})
-  if (sample_inputs, input_tree_spec) in inner_dict:
-    return inner_dict[(sample_inputs, input_tree_spec)]
-  else:
-    hlo = get_xla_computation()
-    _JAX_TO_XLA_COMPUTATION_CACHE[config.trace_context()][jax_func][(
-        sample_inputs, input_tree_spec)] = hlo
-    return hlo
+    tensor_input_flattened = self._flat_input_func.preprocess(args, kwargs)
+
+    abstract_inputs = tuple((a.shape, a.dtype) if a is not None else a
+                            for a in tensor_input_flattened)
+
+    cache_key = hash((abstract_inputs, tuple(self._flat_input_func.non_tensors),
+                      self._flat_input_func.in_spec))
+
+    if cache_key not in self._cache:
+      self._cache[cache_key] = self.especialize(tensor_input_flattened)
+
+    flat_callable = self._cache[cache_key]
+
+    output = flat_callable(tensor_input_flattened)
+    return self._flat_input_func.postprocess(output)
+
+  @abc.abstractmethod
+  def especialize(self, sample_flat_args):
+    pass
+
+
+class JaxCallable(CompiledCallableWithCache):
+
+  def __init__(self, jax_func):
+    super().__init__(FlattenedInputFunc(jax_func))
+
+  def especialize(self, sample_flat_args):
+    import jax
+    tx = maybe_get_torchax()
+    sample_flat_args = tuple(
+        jax.ShapeDtypeStruct(a.shape, tx.ops.mappings.t2j_dtype(a.dtype)
+                            ) if a is not None else None
+        for a in sample_flat_args)
+    with xp.Trace('jax_to_xla_computation'):
+      lowered = jax.jit(
+          self._flat_input_func.flat_call,
+          keep_unused=True).lower(sample_flat_args)
+      hlo_ir = lowered.compiler_ir('hlo')
+      # Get a protobuf representation of the HLO. `as_serialized_hlo_module_proto` is
+      # mentioned at https://github.com/jax-ml/jax/discussions/22266
+      hlo_module = hlo_ir.as_serialized_hlo_module_proto()  # type: ignore
+      computation = computation_from_module_proto('jax_callable', hlo_module)
+      return xla_computation_as_func(computation, 'jax_func')
+
+
+class XlaCallable(CompiledCallableWithCache):
+  """
+  
```

Comment:
"""XlaCallable lets you implement LazyTensor callables using the Python XlaBuilder API."""